apollo-federation/src/query_plan/fetch_dependency_graph.rs

```diff
@@ -2555,15 +3068,579 @@ fn extract_defer_from_operation(
+                path_in_parent: new_path,
+            };
+            dependency_graph.add_parent(*created_node_id, new_parent_relation);
+            created_nodes.insert(*created_node_id);
+        }
+
+        add_post_require_inputs(
+            dependency_graph,
+            fetch_node_path,
+            entity_type_position.clone(),
+            query_graph_edge_id,
+            context,
+            fetch_node_id,
+            new_node_id,
+        )?;
+        created_nodes.extend(new_created_nodes);
+        created_nodes.insert(new_node_id);
+        let initial_fetch_path = create_fetch_initial_path(
+            &dependency_graph.supergraph_schema,
+            &entity_type_position.clone().into(),
+            context,
+        )?;
+        let new_path = fetch_node_path.for_new_key_fetch(initial_fetch_path);
+        Ok((new_node_id, new_path))
+    }
+}
+
+fn paths_has_only_fragments(path: &Arc<OpPath>) -> bool {
+    // JS PORT NOTE: this was checking for FragmentElement which was used for both inline fragments and spreads
+    path.0
+        .iter()
+        .all(|p| matches!(p.as_ref(), OpPathElement::InlineFragment(_)))
+}
+
+fn defer_context_for_conditions(base_context: &DeferContext) -> DeferContext {
+    let mut context = base_context.clone();
+    context.is_part_of_query = false;
+    context.current_defer_ref = base_context.active_defer_ref.clone();
+    context
+}
+
+fn inputs_for_require(
+    fetch_dependency_graph: &mut FetchDependencyGraph,
+    entity_type_position: ObjectTypeDefinitionPosition,
+    query_graph_edge_id: EdgeIndex,
+    context: &OpGraphPathContext,
+    include_key_inputs: bool,
+) -> Result<(SelectionSet, Option<SelectionSet>), FederationError> {
+    // This method is actually called for to handle conditions of @requires, but also to fetch `__typename` in the
+    // case of "fake downcast on an @interfaceObject". In that later case, once we fetched that `__typename`,
+    // we want to wrap the input into the "downcasted" type, not the @interfaceObject one, so that we don't end
+    // up querying some fields in the @interfaceObject subgraph for entities that we know won't match a type
+    // condition of the query.
+    let edge = fetch_dependency_graph
+        .federated_query_graph
+        .edge_weight(query_graph_edge_id)?;
+    let (is_interface_object_down_cast, input_type_name) = match &edge.transition {
+        QueryGraphEdgeTransition::InterfaceObjectFakeDownCast { to_type_name, .. } => {
+            (true, to_type_name.clone())
+        }
+        _ => (false, entity_type_position.type_name.clone()),
+    };
+
+    let Some(edge_conditions) = &edge.conditions else {
+        return Err(FederationError::internal(
+            "Missing edge conditions for @requires",
+        ));
+    };
+
+    let input_type: CompositeTypeDefinitionPosition = fetch_dependency_graph
+        .supergraph_schema
+        .get_type(input_type_name)?
+        .try_into()?;
```

Comment:
We can remap those errors but unsure whether this gives you much more detail than `Schema has no type \"{}\"` (from `get_type`) and `Type "{value}" was unexpectedly not a composite type` (from `try_into`).

Reply:
Thinking about it some more, returning a different error will give us more information about where the error is coming from, since we don't yet have proper diagnostics. So I think we should have this (with `is_composite_type()` instead of individual checks)

```suggestion
    let input_type: CompositeTypeDefinitionPosition = fetch_dependency_graph
        .supergraph_schema
        .get_type(input_type_name)?
        .try_into()?;
    if !input_type.is_object_type()
        || !input_type.is_interface_type()
        || !input_type.is_union_type()
    {
        return Err(FederationError::internal(format!(
            "Type {} should exist in the supergraph and be a composite type",
            &input_type_name
        )));
    }
```